Split out individual difficulties when using difficulty sort at song select

peppy opened the pull request:
Probably the number one feature request for song select. I found a way to implement this without refactoring and rewriting everything, so here's a RFC.

Note that one test is currently broken (pretty obvious why). But everything else seems to work quite well.

Closes https://github.com/ppy/osu/issues/2547 (I guess, until redesign / refactor).

peppy committed:
Split out difficulties in beatmap carousel in a bit of a hacky way

Seems like the simplest path forward for now, without a full rewrite.

peppy committed:
Ensure ID is maintained in temporary `BeatmapSetInfo`s

peppy committed:
Fix incorrect handling when new beatmaps arrive

peppy committed:
Fix selection not being retained when switching between split mode

peppy committed:
Update test to match new behaviour

peppy commented:
Test has been updated, this should be good for review.

peppy committed:
Add test coverage of add/remove when difficulties are split out

peppy committed:
Add coverage of selection retention when difficulties are split out

bdach commented:
Some new intermittent failures here that look potentially relevant in one of the new tests (`TestSceneBeatmapCarousel.TestAddRemoveDifficultySort`):

```
	TearDown : System.TimeoutException : "21 sets visible" timed out: Expected: 21
	  But was:  19
```

```
	TearDown : System.TimeoutException : "21 sets visible" timed out: Expected: 21
	  But was:  20
```

Normally I'd say that looks like "not enough waiting", but `checkVisibleItemCount()` is already supposed to be waiting:

https://github.com/ppy/osu/blob/a64381f8553dd49b9c5c5142aa2934c333b101a4/osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapCarousel.cs#L1213-L1218

peppy commented:
This is likely due to some panels ending off-screen. I'll adjust.

peppy committed:
Limit set/diff count in test to better fit on screen

peppy committed:
Fix one more test probably going off-screen

bdach commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -370,22 +406,48 @@ private void beatmapsChanged(IRealmCollection<BeatmapInfo> sender, ChangeSet? ch
 
-            var newSet = createCarouselSet(beatmapSet);
-            var removedSet = root.RemoveChild(beatmapSet.ID);
+            var removedSets = root.RemoveChild(beatmapSet.ID);
```
The fact that `RemoveChild()` returns multiple items is a bit weird now... That said not sure what would be a better name, either. `RemoveByID()`?

peppy replied:
Renamed to `RemoveItemsByID`.

bdach commented:
Additionally:

Selection is not retained after entering editor and exiting:

https://github.com/ppy/osu/assets/20418176/1e930781-21ef-4771-823f-465e48399f46

Gets even worse if you make changes during:

https://github.com/ppy/osu/assets/20418176/f6fbda3f-3cd4-47fc-85d7-133265cc3663

Possibly update button doesn't appear at all when in split-out mode. Not sure how to actually test this on a real client instance, but doing this in the relevant test scene appears to suggest that it doesn't appear:

```diff
diff --git a/osu.Game.Tests/Visual/SongSelect/TestSceneUpdateBeatmapSetButton.cs b/osu.Game.Tests/Visual/SongSelect/TestSceneUpdateBeatmapSetButton.cs
index 11d55bc0bd..bbb228504b 100644
--- a/osu.Game.Tests/Visual/SongSelect/TestSceneUpdateBeatmapSetButton.cs
+++ b/osu.Game.Tests/Visual/SongSelect/TestSceneUpdateBeatmapSetButton.cs
@@ -15,6 +15,7 @@
 using osu.Game.Overlays.Dialog;
 using osu.Game.Screens.Select;
 using osu.Game.Screens.Select.Carousel;
+using osu.Game.Screens.Select.Filter;
 using osu.Game.Tests.Online;
 using osu.Game.Tests.Resources;
 using osuTK.Input;
@@ -53,10 +54,15 @@ public void SetUpSteps()
         }
 
         [Test]
-        public void TestDownloadToCompletion()
+        public void TestDownloadToCompletion([Values] bool splitOut)
         {
             ArchiveDownloadRequest<IBeatmapSetInfo>? downloadRequest = null;
 
+            if (splitOut)
+            {
+                AddStep("sort by difficulty", () => carousel.Filter(new FilterCriteria { Sort = SortMode.Difficulty }));
+            }
+
             AddStep("update online hash", () =>
             {
                 testBeatmapSetInfo.Beatmaps.First().OnlineMD5Hash = "different hash";

```

peppy commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapCarousel.cs:
```diff
@@ -501,6 +502,39 @@ public void TestAddRemove()
+        [Test]
+        public void TestAddRemoveDifficultySort()
+        {
+            const int local_set_count = 1;
+            const int local_diff_count = 1;
```
Hmm, that's weird. I didn't think I reduced it that low. I think it can be higher than this.

peppy committed:
Fix typo in test step

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -127,14 +129,36 @@ public IEnumerable<BeatmapSetInfo> BeatmapSets
 
         private void loadBeatmapSets(IEnumerable<BeatmapSetInfo> beatmapSets)
         {
+            originalBeatmapSetsDetached = beatmapSets.Detach();
```
The detach was already present in the original code, but was operating on a per beatmap level.. so nothing should have changed. If you're seeing something changed then it requires further investigation.

The idea here is that we only detach once per song select startup (detaching an already detached list should be a no-op).

bdach replied:
Right. Tunnel-visioned a bit there...

peppy committed:
Fix typo in local variable

peppy committed:
Reduce complexity of selection restore

peppy commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -144,6 +168,15 @@ private void loadBeatmapSets(IEnumerable<BeatmapSetInfo> beatmapSets)
                 signalBeatmapsLoaded();
+
+            // Restore selection
+            if (selectedBetmapBefore != null && selectedSetBefore != null && newRoot.BeatmapSetsByID.TryGetValue(selectedSetBefore.BeatmapSet.ID, out var newSelectionCandidates))
```
Nope, good catch.

peppy replied:
Should be better now.

peppy committed:
Adjust test to actually test diff splitting

peppy committed:
Rename remove method to better explain return type being `IEnumerable`

bdach committed:
Add failing test coverage of beatmap update flow w/ split diffs

bdach committed:
Add failing test coverage of selection not being retained on song select

bdach committed:
Fix online ID not being propagated in split difficulty mode

Would result in failures to re-download the beatmap in update flows, for
instance.

bdach committed:
Fix incorrect selection restore code in split case

The fallback to "any of the added sets" needs to be applied after
they've all been added, rather than with every added one. Otherwise, in
flows that expect a particular difficulty to be selected in the end
(such as exiting from editor) would end up switching away from the
edited beatmap.

bdach committed:
Fix changes to beatmap sets being undone on switching sort mode

bdach committed:
Merge branch 'master' into carousel-difficulty-split

bdach commented:
I've fixed a few remaining things that I would consider issues. There are a few wonky spots still in how the "update beatmap" button or the "add to collection" context menu interact with this feature but they're not outright bugs so it's probably ok to let them be for now.

@peppy please double check that you're OK with my changes. Probably better to let CI run on this one too because I didn't run everything on my machine (just the most relevant ones), and there are new tests here so let's see how the ~~fuzzers~~ github runners do on them.

bdach committed:
Fix `TestSceneUpdateBeatmapSetButton` using random difficulty count

peppy committed:
Merge branch 'master' into carousel-difficulty-split

peppy commented:
Changes look solid, thanks 👍 .
